index.rb

```diff
@@ -1,160 +1,250 @@
 
-  def my_count(*args)
-    if args.length == 0 then
-      return puts self.length
+    if args.length == 2 && !block_given?
+      my_aggregator = args.first
+
+      each do |i|
+        my_aggregator = my_aggregator.send(args[1], i)
+      end
+
+      return my_aggregator
     end
 
-    my_parameter = args.first
-    my_count_counter = 0
-    for i in self
-      if(i == my_parameter) then
-        my_count_counter += 1
+    if args.length == 1
+      my_aggregator = args.first
+
+      each do |i|
+        my_aggregator = yield(my_aggregator, i)
       end
+
+      return my_aggregator
     end
-    puts my_count_counter
-  end
 
-  def my_map
+    my_aggregator = first
 
-    my_array = []
-    
-    #Replace the .each with my_each
+    each do |i|
+      next if i == first
 
-    for i in self
-      my_array.push(yield(i))
+      my_aggregator = yield(my_aggregator, i)
     end
-    my_array
 
+    my_aggregator
   end
 
+  # rubocop:enable Metrics/MethodLength
+
+  def my_count(*args)
+    if args.length.zero? && !block_given?
+      return size if is_a? Range
 
-  def my_inject
+      return length
+    end
 
-    my_aggregator = self.first
-    
-    for i in self do
-      if i == self.first then
-        next
+    if block_given?
+      my_count_counter = 0
+      each do |i|
+        my_count_counter += 1 if yield(i)
       end
-      my_aggregator = yield(my_aggregator, i)
+      return my_count_counter
     end
 
-    puts my_aggregator
-
+    my_parameter = args.first
+    my_count_counter = 0
+    each do |i|
+      my_count_counter += 1 if i == my_parameter
+    end
+    my_count_counter
   end
 
-  def self.multiply_els(my_array)
+  def my_map_prock(block)
+    my_array = []
 
-    my_array.my_inject {|a,b| a * b}
+    each do |i|
+      my_array.push(block.call(i))
+    end
+    my_array
   end
 
+  def my_map(block = nil)
+    return to_enum(:my_map) unless block_given?
+    return my_map_prock(block) unless block.nil?
 
-end
-
-
-#Testing all the codes,  ADD YOUR TESTS HERE
-[1,2,3].my_each { |n| puts n * 3}
-[1,2,3].my_each_with_index { |number, index| puts "the number: #{number}, has index: #{index}"}
-[1,2,3,4].my_select { |n| n.even?}
-[0,2,4,8].my_all? {|n| n.even?}
-[1,3,5,7].my_any? {|n| n.even?}
-[1,3,5,7].my_none? {|n| n.even?}
-[1,3,5,7,1,nil,5,7,1,nil,5,7].my_count
-[1,3,5,7,1,nil,5,7,1,nil,5,7].my_count(1)
-some_new_array = [1,3,4,8].my_map { |n| n * 3}
-some_other_new_array = [1,3,4,8].my_each {|n| n  * 3}
-puts some_new_array
-puts some_other_new_array
-[1,3,5,7].my_inject { |a,b| a + b}
-
+    my_array = []
 
-movie = Proc.new {return "something"}
-puts movie.call
+    each do |i|
+      my_array.push(yield(i))
+    end
+    my_array
+  end
 
-Enumerable.multiply_els([1,3,4,7,8,12])
+  def multiply_els(*args)
+    return args.first.my_inject { |a, b| a * b } if args.length == 1
 
+    my_inject { |a, b| a * b }
+  end
```

Comment:
- [ ] This should not be inside the enumerable block. It should be a standalone method.